Package highlighting and package colors with gradient

mrpatpat opened the pull request:
# Package highlighting and package colors with gradient

## Description

This PR implements #152 and #155. It also allows #165. Please keep in mind that the tests are still missing. I will add them before resolving this PR. The purpose of this PR is to get feedback and reviews for the current state. 

### Context menu
The context menu opens with a right click on any node (map or treeview). It contains a few convenience actions like a visibility toggle or isolate option.
It also offers a color selection when a package was right clicked. These colors are the highlighting colors. If you click an already highlighted package then a transparent color appears in the selection. It removes the highlighting. 

![capture](https://user-images.githubusercontent.com/2622069/42032157-f8e7a80c-7ad8-11e8-96fb-cff2f0a5ed0a.PNG)

![capture](https://user-images.githubusercontent.com/2622069/42032193-1223279c-7ad9-11e8-9ac2-6d574eb9e30d.PNG)

![capture](https://user-images.githubusercontent.com/2622069/42032235-2df2f9ca-7ad9-11e8-910b-b577ea121a30.PNG)

### Highlighting
Highlighting a package tints the package and its descendants with the given color. If one of the descendents is already tinted, then this descendant and its children will be ignored. This prevents overwritting of already highlighted packages. 
De-Highlighting will also ignore other colors than the color of the selected node. It therefore removes only one color at a time. 

![capture](https://user-images.githubusercontent.com/2622069/42032297-6c71c6ea-7ad9-11e8-82fd-5fd72c7b0cfb.PNG)

To increase usability the folder icons of the tree view are also colored in their highlighting color. 

![capture](https://user-images.githubusercontent.com/2622069/42032382-b529642e-7ad9-11e8-8e21-45dbe0f70713.PNG)

### Floor color

A discussion regarding the floor color can be found in #152. This is still an open topic.

![capture](https://user-images.githubusercontent.com/2622069/42032478-f7a0e5c0-7ad9-11e8-89f7-714b316365e0.PNG)

### cc.json

This PR also allows #165 since the loading service does not use the json schema restrictively e.g. It does not forbid unmentioned properties. It is possible to give a node the property "markingColor": "0xff0000" to make it red if it is a folder. #165 should decide if this behaviour should be allowed or forbidden.

### Code map actions service

Part of this PR is a new service which bundles actions on the currently viewed code map e.g. hiding nodes, highlighting them etc. 

### Treemap service

Since this PR is just the first one in a series of features depending on the treemap service, I decided to clean this service up and introduce the correct types. 

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

* [x] **All** requirements mentioned in the issue are implemented
* [x] Does match the Code of Conduct and the Contribution file 
* [x] Task has its own **GitHub issue** (something it is solving)
  * [x] Issue number is **included in the commit messages** for traceability
* [x] **Update the README.md** with any changes/additions made
* [x] **Update the CHANGELOG.md** with any changes/additions made
* [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
* [x] **All tests pass**    
* [x] **Descriptive pull request text**, answering:
  + What problem/issue are you fixing?
  + What does this PR implement and how? 
* [x] **Assign your PR to someone** for a code review
  + This person _will be contacted **first**_ if a bug is introduced into `master`
* [x] **Manual testing** did not fail

felixherbst commented:
The "clear color" icon is not recognizable as the white box it currently is. Instead of a "remove" button, there could just be the old color field with a remove sign. 
![image](https://user-images.githubusercontent.com/8899664/42084407-d05d046c-7b8d-11e8-97e3-ad0ce070117d.png)


Colors are still alternating when bases are highlighted.
![image](https://user-images.githubusercontent.com/8899664/42089664-91d351ae-7b9e-11e8-963b-4f51d6099dbb.png)

alschmut commented:
I really like highlighting the packages and the changes on this branch! Also the refactored treemap.service looks a lot cleaner 👍
- But we could align the darker alternating package colors to a similar lighter color tone.
- The context menu always jumps from the old position to the new one. Did you already try to solve this?
- When opening the context menu on a building we could add the option to highlight the underlying (parent node) package?!
- Do we need to add a "reset all highlightings" button? (also possible in future issue)
- And i agree with the comment of @felixherbst

mrpatpat commented:
Thank you for your contributions. I adjusted the color alternation and made it more subtle for marking purposes. 
I also added a cross over the color of the current node. Clicking it will remove the color. 

![capture](https://user-images.githubusercontent.com/2622069/42516217-d2e13856-845d-11e8-94c3-bc563e602022.PNG)

I recommend to create feature requests for these two suggestions:

> When opening the context menu on a building we could add the option to highlight the underlying (parent node) package?!

> Do we need to add a "reset all highlightings" button? (also possible in future issue)
